### ISSUE TITLE
testing a dotnet core class library with Specflow 

### ISSUE THREAD

**Issue by jimdeane** [stajs/SpecFlow.NetCore#46]
I've been trying to understand what this means for a couple of days now. My understanding (probably totally wrong) was that i could use VS2017 to develop a DOTNET CORE class library and then test it using another DOTNET CORE Test Project, using the SpecFlow.NetCore NuGet package (and modifications to the app.config).

The behaviour appears conflicting.

The .feature.cs gets generated but the Techtalk.Specflow library is missing.

If I include SpecFlow 2.1.0 the build fails with incompatible .net versions.

Am I simply trying to use this in an impossible way or is there a workaround.

We use Specflow universally and a attempting to start a DOTNET CORE project right now and failing miserable.

Thnaks

**Comment by igor-1024** [comment]
Hi Jim! I had exactly the same problems: “.feature.cs” gets generated with TechTalk references or I get “incompatible .NET versions” error. I ended up moving all of the logic out into a separate .NETStandard 1.4 class library, referenced it by original .NETCoreApp 1.1 project and had .NET 4.6.1 SpecFlow Test project testing just the business logic in 1.4 class library the “old” way.

**Comment by stajs** [comment]
As [@erik-lundgren mentioned](https://github.com/stajs/SpecFlow.NetCore/issues/45#issuecomment-290456988), it is a case of SpecFlow still being stuck on full framework (see #39).

@igor-1024's suggestion is the best workaround for now.

**Comment by tomitrescak** [comment]
trying to understang @igor-1024 's approach. How exactly do you refrence .NET core projects from NET4.6? I tried but it's not doable.

**Comment by smudge202** [comment]
Not the right place to ask, @tomitrescak but the statement itself is slightly off.

You can test a `netstandard2.0` library from a `net462` test project, and the same library can be used by a `netcoreapp2.0`.

This has become much more confused than it should have been thanks to a ~shit storm~ bad decision MS made. Long story short:

MS wanted to introduce new APIs for ASP.Net Core. The obvious (and originally intended) thing to do would have been to create the next version of .Net Standard containing the APIs. Then they could continue writing libraries such as MVC Core against the .Net Standard. However, that would give us a version of .Net Standard that the full framework (*platform*) could not implement until the next full release of .Net.

So, they chickened out. They added the new APIs only to the `netcoreapp` (platform) and then, as of 2.0, changed all the libraries to also point to the `netcoreapp` *platform* instead of `netstandard`.

This makes life very difficult, because libraries that want to interact with MVC components are then forced off of `netstandard` and onto `netcoreapp`. So, if you have a library that is being forced into targeting `netcoreapp`, it is *not* possible to test it using `net462`. Bear in mind though, despite MS trying to hide this blunder by defaulting the class library template to `netcoreapp`, the first thing I do every time is correct this and change it to `netstandard2.0` and do anything I can to keep it that way.

Hopefully that explains why some of this is in the state it is. We need one of two remedies:
* MS to fix their mistake, increment .Net Standard and then hang about until `net463` or whatever which implements the new standard
* SpecFlow to properly support `netcoreapp`

Neither of which are possible from this repo. 😞

**Comment by smudge202** [comment]
> My original reply was wrong so have replaced it with this:

@tomitrescak by default, your class libraries target `netcoreapp2.0` (this is a mistake in the templates if you ask me). However, most of the net core ecosystem (MVC for example), if you check out their dependencies [on NuGet](), actually target `netstandard2.0`. If you change your class library to also target `netstandard2.0` (edit the csproj) then both the `net462` and `netcoreapp2.0` platforms are able to reference the library.

There is a downside to doing this however - the platforms implement `netstandard` differently, and there *could* be a difference in behaviour at times. MS strive to avoid this but don't always succeed. Therefore, if you test a `netstandard` library against `net462` in order to allow specflow to run, but in production you run the app using the `netcoreapp` platform, there *could* (though shouldn't) be slight differences.

Unfortunately, if you want SpecFlow to be involved, there's no real way around this right now. Ideally, your test project would *multi-target* both `net462` and `netcoreapp2.0` in order to test against both platforms (this kind of thing isn't even supported by the VS Test Explorer yet, but is something both supported by and preached for xUnit / Brad Wilson).

TL;DR: Try targeting your class library at `netstandard2.0` (hopefully your nuget dependencies all support this). If the code your testing *is* an application, move the code into a class library and test it there.
